tests/solidity_test_helpers/ZeroX/ZeroXPoC.sol

```diff
@@ -0,0 +1,446 @@
+import 'Augur.sol';
+import 'IController.sol';
+
+contract ZeroXPoC is ReentrancyGuard {
+    using SafeMathUint256 for uint256;
+
+    // Error Codes
+    enum Errors {
+        ORDER_EXPIRED,                    // Order has already expired
+        ORDER_FULLY_FILLED_OR_CANCELLED,  // Order has already been fully filled or cancelled
+        ROUNDING_ERROR_TOO_LARGE,         // Rounding error too large
+        INSUFFICIENT_BALANCE_OR_ALLOWANCE // Insufficient balance or allowance for token transfer
+    }
+
+    uint16 constant public EXTERNAL_QUERY_GAS_LIMIT = 4999;    // Changes to state require at least 5000 gas
+
+    // token => account => balance
+    mapping(address => mapping(address => uint256)) public tokenBalances;
+
+    // orderHash => amounts of Amount filled or cancelled.
+    mapping (bytes32 => uint) public filled;
+    mapping (bytes32 => uint) public cancelled;
+
+    Augur public augur;
+    CompleteSets public completeSets;
+    IController public controller;
+    ICash public cash;
+
+    event Fill(
+        address taker,
+        uint256 amountFilled,
+        bytes32 orderHash
+    );
+
+    event Cancel(
+        bytes32 orderHash,
+        uint256 cancelledAmount
+    );
+
+    event Error(
+        uint8 indexed errorId,
+        bytes32 indexed orderHash
+    );
+
+    struct Order {
+        address maker;
+        IMarket market;
+        uint256 outcome;
+        uint256 orderType;
+        uint256 amount;
+        uint256 price;
+        uint expirationTimestampInSec;
+        bytes32 orderHash;
+    }
+
+    event Deposit(
+        address indexed account,
+        address indexed shareToken,
+        uint256 amountDeposited,
+        uint256 amountHeld
+    );
+
+    event Withdraw(
+        address indexed account,
+        address indexed shareToken,
+        uint256 amountWithdrawn,
+        uint256 amountHeld
+    );
+
+    function ZeroXPoC(Augur _augur) public {
+        augur = _augur;
+        controller = _augur.getController();
+        completeSets = CompleteSets(controller.lookup("CompleteSets"));
+        cash = ICash(controller.lookup("Cash"));
+        cash.approve(augur, 2**254);
+    }
+
+    /*
+    / Market Share management
+    */
+
+    function deposit(ERC20 _token, uint256 _amount) public nonReentrant returns (bool) {
+        require(_token != ERC20(0));
+        require(_token.transferFrom(msg.sender, this, _amount));
+        tokenBalances[_token][msg.sender] = tokenBalances[_token][msg.sender].add(_amount);
+        Deposit(msg.sender, _token, _amount, tokenBalances[_token][msg.sender]);
+        return true;
+    }
+
+    function withdraw(ERC20 _token, uint256 _amount) public nonReentrant returns (bool) {
+        require(_token != ERC20(0));
+        uint256 _heldAmount = tokenBalances[_token][msg.sender];
+        require(_heldAmount >= _amount);
+        require(_token.transfer(msg.sender, _amount));
+        tokenBalances[_token][msg.sender] = _heldAmount.sub(_amount);
```

Comment:
Move the balance modifications above external calls? (even though it is modified to nonReentrant)